jmbitcoin/jmbitcoin/__init__.py

```diff
@@ -1,4 +1,17 @@
 import coincurve as secp256k1
+
+# If libsecp is available already, use it
+# unaltered; else we use the version installed
+# via Joinmarket's install.sh script
+# See: https://github.com/Simplexum/python-bitcointx/commit/79333106eeb55841df2935781646369b186d99f7#diff-1ea6586127522e62d109ec5893a18850R301-R310
+import os, sys
+try:
+    import bitcointx.core
+except ImportError:
+    import bitcointx
+    bitcointx.set_custom_secp256k1_path(os.path.join(
+        sys.prefix, "lib", "libsecp256k1.so"))
```

Comment:
Wouldn't it be more robust to always set the path to your supplied lib ? Of course libsecp256k1 is fairly stable and most likely the one in the default path would be OK, but if you are supplying the library anyway, I'd always use it, at least to make any related errors/problems more deterministic.

Reply:
Yes, I wavered on this point. I wondered if there might be scenarios where the install had secp compilation be optional. Otoh there's what you said, which worried me: what if the system-installed version is somehow incompatible.
I suppose the second point might be more important than the former. I'm honestly not sure.

Reply:
In my opinion, the optionality of "install and secp compilation" is the decisive factor. If there is always an installation-supplied lib, using it make the behavior more deterministic, which is good.

One way I see is to have both options by prefer the installation-supplied lib first. First call `load_secp256k1_library(local_path)` with an installation-specific path, and then if there's an error, try `load_secp256k1_library(None)` to see if there is system-supplied lib. If there's no error with `local_path`, then call `set_custom_secp256k1_path(local_path)`, otherwise do not call `sec_custom..` and the system-supplied lib will be used.

Reply:
Hmm, it seems that it would not be possible to call `load_secp256k1_library` without `_secp256k1` global var of `secp256k1.py` module to be set. So it will need to be a call to `ctypes.cdll.LoadLibrary()` instead of `load_secp256k1_library`:

```
    try:
        handle = ctypes.cdll.LoadLibrary(path)
    except Exception as e:
        ... can't find the lib ...
 ```

When `path` is None this means to search for system-installed library.

This is more low-level than calling functions from `bitcointx`, but this is probably not a big issue, because the problem that this solves _is_ on a low-level.

Reply:
Alternatively, there might be no need to actually _try_ to load the installation-supplied libsecp library, and it might be enough to just check for existence of the file. If the file exists, but cannot be loaded as a library, this is a problem with installation that has to be fixed, rather than ignored by falling back to system-supplied library.

Reply:
> not possible to call load_secp256k1_library without _secp256k1 global var of secp256k1.py module to be set.

because `load_secp256k1_library()` is defined in `bitcointx.core.secp256k1.py`, and `_secp256k1` global handle is set right there at the end of the module, on first load.

Reply:
> Alternatively, there might be no need to actually try to load the installation-supplied libsecp library, and it might be enough to just check for existence of the file.

Ah, very good point. There's a big difference between "try to load a shared object you think should be on the system" and "try to load a shared object you have installed yourself". In the latter case yeah just check if the file is there. So then I lean towards your alternative algo: if JM installed lib exists use it, else try to load using system. I will rewrite the commit on that basis.

Reply:
I would also try to import `bitcointx.core.secp256k1` instead of just `bitcointx.core`, because the dependency of `bitcointx.core` on the `secp256k1` is indirect via `script.py` -> `key.py` -> `secp256k1.py`. It is unlikely that `bitcointx.core` would ever *not* depend on `secp256k1`, but you never know what can happen in the future..

Reply:
I just went for the file check as per https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/536/commits/0151b1fb5239c9bead0216b89c1c826aa421ff47